src/main/java/de/hpi/bpt/argos/storage/hierarchy/HierarchyBuilderImpl.java

```diff
@@ -0,0 +1,89 @@
+package de.hpi.bpt.argos.storage.hierarchy;
+
+import de.hpi.bpt.argos.core.Argos;
+import de.hpi.bpt.argos.storage.PersistenceAdapterImpl;
+import de.hpi.bpt.argos.storage.dataModel.entity.Entity;
+import de.hpi.bpt.argos.storage.dataModel.entity.VirtualRoot;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * {@inheritDoc}
+ * This is the implementation.
+ */
+public final class HierarchyBuilderImpl implements HierarchyBuilder {
+
+	private EntityHierarchyNode rootNode;
+
+	private static HierarchyBuilder instance;
+
+	/**
+	 * This constructor hides the default public one to implement the singleton pattern.
+	 */
+	private HierarchyBuilderImpl() {
+
+	}
+
+	/**
+	 * This method returns the singleton instance of this class.
+	 * @return - the singleton instance of this class
+	 */
+	public static HierarchyBuilder getInstance() {
+		if (instance == null) {
+			instance = new HierarchyBuilderImpl();
+		}
+
+		return instance;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public EntityHierarchyNode getEntityHierarchy() {
+		if (rootNode != null && !Argos.getTestMode()) {
+			return rootNode;
+		}
+
+		List<Entity> entities = PersistenceAdapterImpl.getInstance().getEntities();
+		rootNode = new EntityHierarchyNodeImpl(VirtualRoot.getInstance().getId());
+
+		addChildren(rootNode, entities);
+
+		return rootNode;
+	}
+
+	/**
+	 * This method adds all child entities to a parent entity.
+	 * @param parent - the parent
+	 * @param listOfAllEntities - a list of all remaining entities
+	 * @return - a list fo all remaining entities after they have been added to a parent node
+	 */
+	private List<Entity> addChildren(EntityHierarchyNode parent, List<Entity> listOfAllEntities) {
+		if (listOfAllEntities.isEmpty()) {
+			return listOfAllEntities;
+		}
+
+		List<Entity> entities = new ArrayList<>(listOfAllEntities);
+		List<EntityHierarchyNode> children = new ArrayList<>();
+
+		for (Iterator<Entity> it = entities.iterator(); it.hasNext();) {
+			Entity child = it.next();
+			if (child.getParentId() != parent.getId()) {
+				continue;
+			}
+			EntityHierarchyNode newChild = new EntityHierarchyNodeImpl(child.getId());
+			children.add(newChild);
+			parent.addChild(newChild);
+			it.remove();
+		}
+
+		for (EntityHierarchyNode child : children) {
+			entities = addChildren(child, entities);
```

Comment:
Entities is gonna be empty anyways and you never use the returned value in your other methods, so why letting it return anything at all and also why overrwriting the entities variable if it's gonna be empty and unused anyway?